Add Light class pythonizing vtkLight

adeak opened the pull request:
### Overview

This PR adds a `pyvista.Light` class which is a pythonic subclass of `vtk.vtkLight`. Originally suggested in https://github.com/pyvista/pyvista/issues/974.

### Details

Main features:

  1. `Light` is a `vtk.vtkLight` subclass, pyvista objects return `pyvista.Light`s now. A `pyvista.Light.from_vtk` classmethod facilitates converting from base class instances (by copying).
  2. New `LightType` enum (class "constants") with pretty printing.
  3. `add_light()`, `remove_all_lights()` on `Plotter`s and `Renderer`s. For the former class a new kwarg specifies if each renderer should be affected or only the active one. Also new `lighting` `__init__` kwarg for `Plotter`s, in order to enable starting a plotter without any lights.

My example in https://github.com/pyvista/pyvista/issues/974#issuecomment-725972305 now looks like this:
```py
import pyvista as pv
from pyvista.examples import download_nefertiti

data = download_nefertiti()

plotter = pv.Plotter(lighting=None)
plotter.add_mesh(data, color='white', smooth_shading=True)
plotter.set_background('black')

light_positions = [(-2, 2, 0), (2, -2, 0)]
light_colors = [(1, 0, 0), (0, 0, 1)]

for light_pos, light_color in zip(light_positions, light_colors):
    light = pv.Light(position=light_pos, color=light_color)
    light.focal_point = (0, 0, 0)
    
    plotter.add_light(light)

plotter.show(cpos=[(-600, -850, 200), (0, 0, 0), (0, 0, 1)], screenshot='nefertiti_lit_upgrade.png')
```
(although to be fair the cone angle setting was redundant in my original snippet).

---

The lights and the light types have pretty `repr`s:
```py
>>> import pyvista as pv
>>> plotter = pv.Plotter()  # default: `vtkLightKit` with 5 lights, converted to `pyvista.Light`s
>>> plotter.renderer.lights
[<pyvista.Light (Headlight) at 0x7ffa114f96a0>, <pyvista.Light (Camera Light) at 0x7ffa114f98e0>, <pyvista.Light (Camera Light) at 0x7ffa114f9940>, <pyvista.Light (Camera Light) at 0x7ffa114f99a0>, <pyvista.Light (Camera Light) at 0x7ffa114f9a00>]
>>> light = plotter.renderer.lights[0]
>>> light.light_type
<LightType.HEADLIGHT: 1>
>>> light.light_type == light.HEADLIGHT == 1
True
>>> print(f'This light is a {light.light_type}.')
This light is a Headlight.
```

---

`Plotters` have a new `__init__` kwarg `lighting` which offers some customization:
   1. `lighting='light kit'`: what used to be the default. In order to ensure that every light we create is a `pyvista.Light`, a `vtkLightKit` is temporarily created, then converted to individual `pyvista.Lights` using the renderer as a parser. (This is the only point where wrapping `vtkLight` with composition would be superior to inheriting from it.)
   2. `lighting='three lights'`: calls an updated version of `plotter.enable_3_lights()`, which now only spawns the 3 necessary lights that would stay turned on. (In the original implementation the `vtkLightKit`'s five component lights are repurposed for 3-lights illumination.)
   3. `lighting='none'` (or `None`): does not add any lights to the renderer, anticipating manual lights to be added later (see my above example).

`Plotter.add_light` and `Plotter_remove_all_lights` delegate the work to their respective renderer, either to the active renderer or their every existing renderer.

### TODOs

  - [x] finalize the API
  - [x] clean up dozens of TODOs
  - [x] add a lot more input validation and corresponding tests
  - [x] add doc section and examples


Resolves https://github.com/pyvista/pyvista/issues/974.

adeak commented:
@akaszynski I would like to take you up on your offer for guidance implementing this PR. Before I write examples and doc sections I'd like to make sure we're on the same page regarding the API.

Non-trivial decisions or ambiguities I'd like feedback on (I have a bunch of TODOs scattered in the code right now; these are the major or uncertain ones):

  1. Should I implement a `light.color = ...` with `setattr`? A property is not appropriate unless we link all three kinds of colours (unlike vtk behaviour). Also linked to the fate of `set_color`. The problem is that `vtkLight.SetColor` only sets two out of three colour types (diffuse and specular yes, ambient no). It would be nice to set all three in pyvista, but diverging from vtk behaviour might be an issue.
  2. Can a transformation matrix appear accidentally for a light? I have no idea. Do we want to expose setting transformation matrices? If the answer to either is "yes": do the property names `world_position`, `world_focal_point` make sense? If transformation matrices can only appear by setting them manually, we might not expose these at all, for now.
  3. Is `is_on` an OK name for the property of the switch state? `.switch` sounds unintuitive, `.on` sounds too short.
  4. Is it OK to call a deepcopy a "copy"? I haven't implemented this yet. `vtkLight`s can do shallow copies, which as I understand means the lights share the same transformation matrix. This might be a moot point depending on point 2. above.
  5. Plotter's lighty methods (`add_light`, `remove_all_lights`): now with `only_active` kwarg, specifying whether only the active renderer should be affected, or all of them. To me the default should be "all of them" (and this is how it's implemented right now), but existing behaviour (mainly in `disable_3_lights`) was "only affect active renderer"...
  6. I've replaced the `vtkLightKit` default with `pyvista.Light` lights. This is necessary to ensure that `renderer.lights` gives us `pyvista.Light` objects. But then should we add two placeholder lights in case of "3 lights" illumination for backwards compatibility? From a github search only pyvista code calls `enable_3_lights()`... The current implementation scraps all existing lights and only adds the 3 necessary lights for "3 lights" illumination.
  7. Do we need @skip_no_plotting decorator for tests when there's no plotters, only lights?
  8. There might be a bit too many things called `lighting` now:
    1. `Plotter` init (specifying lights on creation)
    2. `Plotter.add_mesh` kwarg (specifying if the given mesh should be lit; defaults to `rcParams['lighting']`)
    3. `Plotter.lighting` which used to be the `vtkLight` active by default...
    It might be necessary to rename at least the new init kwarg. But the `.lighting` attribute makes little sense now that the `vtkLightKit` is replaced by individual `pyvista.Light` objects.

JevinJ commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,465 @@
+
+    # TODO: better/more explanation for ``position``?
+
+    # pull in light type enum values as class constants
+    HEADLIGHT = LightType.HEADLIGHT
+    CAMERA_LIGHT = LightType.CAMERA_LIGHT
+    SCENE_LIGHT = LightType.SCENE_LIGHT
+    # TODO: is this design OK?
```
I would just use the enum directly.

adeak replied:
The reason I wanted to pull these in into the class is because I can easily imagine use cases where we want to compare light types from outside the `Light` class. Say, if I end up with a list of light types for some reason. In this case I'd want to be able to do `pyvista.Light.HEADLIGHT`. I guess we could also pull in `LightType` into the pyvista namespace, but to me that would seem overkill (it's not important enough to pollute the main namespace with it).

JevinJ commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,465 @@
+        """
+        self.SetLightTypeToSceneLight()
+
+    @classmethod
+    def from_vtk(cls, vtk_light):
+        """Create a Light object from a vtkLight, resulting in a copy.
+
+        Parameters
+        ----------
+        vtk_light : vtkLight
+            The ``vtkLight`` to be copied.
+
+        """
+        if not isinstance(vtk_light, vtkLight):
+            raise TypeError(f'Expected ``vtkLight`` object, got ``{type(vtk_light)}`` instead.')
+
+        light = cls()
+        light.light_type = vtk_light.GetLightType()  # resets transformation matrix!
+        light.position = vtk_light.GetPosition()
+        light.focal_point = vtk_light.GetFocalPoint()
+        light.ambient_color = vtk_light.GetAmbientColor()
+        light.diffuse_color = vtk_light.GetDiffuseColor()
+        light.specular_color = vtk_light.GetSpecularColor()
+        light.intensity = vtk_light.GetIntensity()
+        light.is_on = vtk_light.GetSwitch()
+        light.positional = vtk_light.GetPositional()
+        light.exponent = vtk_light.GetExponent()
+        light.cone_angle = vtk_light.GetConeAngle()
+        light.attenuation_values = vtk_light.GetAttenuationValues()
+        light.shadow_attenuation = vtk_light.GetShadowAttenuation()
+        # TODO: copy transformation matrix even if not exposed?
+
+        return light
```
I read from bane(I think) that pyvista objects should just handle what is thrown at them from the constructor, instead of classmethods. Though I disagree, this isn't how the others are constructed currently.

adeak replied:
And I absolutely agree! But I opted not to do that exactly because I saw that the rest of the API is not designed that way :) My preference would be to have a single optional positional-only `vtkLight` argument, and everything else keyword-only arguments.

JevinJ commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,465 @@
+        if position:
+            self.position = position
+
+        if color:
+            self.ambient_color = color
+            self.diffuse_color = color
+            self.specular_color = color
+            # TODO: make this a single call if that's implemented
```
Do they have to be the same color? Could we default to white or black for each, or does that cause issues?

adeak replied:
They don't _have_ to be, which is why we can set them separately. But if a single "color" is passed the only reasonable thing (to me) is to set every component's color to this.

Our default preserves `vtkLight`'s default, which is black for ambient and white for specular and diffuse. This seems to match a `.SetColor` call with white: that would only set specular and diffuse. If there's reason to leave out ambient unless explicitly requested (as `vtkLight` does it) I can change this too.

As for the TODO comment: I just meant that if `set_color` is changed to set all 3 then I can replace those 3 lines with a single call to `set_color`. We'd still have access to the individual component colour properties.

JevinJ commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,465 @@
+            # be forgiving: ignore spaces and case
+            light_type_orig = light_type
+            type_normalized = light_type.replace(' ', '').lower()
+            mapping = {'headlight': LightType.HEADLIGHT,
+                       'cameralight': LightType.CAMERA_LIGHT,
+                       'scenelight': LightType.SCENE_LIGHT,
+                      }
+            light_type = mapping.get(type_normalized)
+            if light_type is None:
+                raise ValueError(f'Invalid ``light_type`` "{light_type_orig}"')
```
How about something like(not tested):

```suggestion
            # be forgiving: ignore spaces and case
            type_normalized = light_type.replace(' ', '').lower()
            mapping = {str(e).replace(' ', '').lower(): e for e in LightType}
            try:
                light_type = mapping[type_normalized]
            except KeyError:
                raise ValueError(f'Invalid ``light_type`` "{light_type}"')
```

adeak replied:
I thought about automating it with something like that, but I was worried that it might make the intention harder to read and so the code harder to maintain.

As for LBYL vs EAFP: well, both can work ;) Not sure which one is more prevalent in the codebase, I'll try to see if there are other examples.

JevinJ commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,465 @@
+
+    def __repr__(self):
+        """Print a repr specifying the id of the light and the vtkLight it wraps."""
+        return (f'<pyvista.Light ({self.light_type}) at {hex(id(self))}>')
```
```suggestion
        return (f'<{self.__class__.__name__} ({self.light_type}) at {hex(id(self))}>')
```

adeak replied:
Good catch, thanks! I originally implemented the `repr` for a composition-based approach. In this case I would have stressed that this is a `pyvista.Light` which wraps a `vtkLight`. This is no longer needed because we aren't wrapping anything.

Also, you can see the obsolete docstring for the `__repr__` which mentions a `vtkLight` being wrapped. Obviously this is nonsense now. Thanks.

JevinJ commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,465 @@
+        if not isinstance(ltype, int):
+            raise TypeError('Light type must be an integer subclass,'
+                            f' got {ltype} instead.')
+        # LightType is an int subclass; convert to proper int
+        ltype = int(ltype)
+        self.SetLightType(ltype)
```
Comment isn't super helpful.

```suggestion
        self.SetLightType(int(ltype))
```

adeak replied:
I kind of agree, but I wanted to make it clear that `LightType` values are covered by the `isinstance` check. If you think that's trivial then no comment is needed. (Or I'm open to suggestions for a more helpful comment.)

But now I realize that exactly because of the `int` subtype we can directly feed the type to `SetLightType`, so the `int(ltype)` call itself is superfluous.

JevinJ replied:
Specifying that LightType is accepted in the docstring might be more helpful, and if vtk accepts it without casting then it doesn't matter IMO.

adeak replied:
I'm mostly worried about devs having to spend mental CPU cycles on figuring out that the `int` case covers `LightType`. The docstring is more for end users in my head. But anyway I've mentioned these enum values in the `Light` class docstring. When I get to writing the proper documentation and examples I'll take care to make this clear.

But you're probably right about it not being an issue without an explicit `int` conversion. Thanks.

akaszynski commented:
> Instead I want to suppress the context altogether, because as far as the user is concerned they only need to see the final error:

Agreed.  Sometimes a full traceback and detailed error is helpful, but I think it's good practice to use `raise from` to produce concise errors for the user.

JevinJ commented:
I've improved some of the tests with Hypothesis and added clamping of some inputs on your fork.

https://github.com/adeak/pyvista/pull/1

adeak commented:
@JevinJ thanks. A few remarks:

  1. The PR might seem stalled, but really I'm just waiting for feedback concerning the design (potentially in the form of "_yeah, it's fine, go for it_"). As I noted above there's plenty of input validation to be done, I intentionally put these off until we're sure that these are the methods to be validated.
  2. I don't agree with clamping. If the user sets nonsense data we should either raise, or pass it on to vtk which probably clamps internally (I haven't tested yet). A colour value larger than 1 is probably wrong, and might actually be the result of, say, uint8 values.
  3. I've never run into hypothesis, including pyvista (it doesn't seem too prevalent in the codebase). I'll have to read into it and understand how it works before I pull it into my work.

JevinJ commented:
> @JevinJ thanks. A few remarks:
> 
>     1. The PR might seem stalled, but really I'm just waiting for feedback concerning the design (potentially in the form of "_yeah, it's fine, go for it_"). As I noted above there's plenty of input validation to be done, I intentionally put these off until we're sure that these are the methods to be validated.
> 
>     2. I don't agree with clamping. If the user sets nonsense data we should either raise, or pass it on to vtk which probably clamps internally (I haven't tested yet). A colour value larger than 1 is probably wrong, and might actually be the result of, say, uint8 values.
> 
>     3. I've never run into hypothesis, including pyvista (it doesn't seem too prevalent in the codebase). I'll have to read into it and understand how it works before I pull it into my work.

2. Vtk does not clamp internally, I ran the tests before adding the logic. I honestly think it's better to just handle slightly wrong input if it can be coerced into being correct, maybe show a warning too, but if it's truly garbage then raise. But it's up to you/@akazinsky I suppose.

3. I added hypothesis not too long ago, it's pretty cool but I seem to be the only one who uses it. Spend less time worrying about having full example coverage and more time writing better tests.

adeak commented:
Sorry, when I said "clamp" I meant that it just handles outlying values as if they were min/max. Is that also what you meant? I know that setting an intensity of 2 will be preserved as-is. What I didn't check is whether this would lead to the same behaviour as if the intensity were 1 (which is my guess).

JevinJ commented:
> Sorry, when I said "clamp" I meant that it just handles outlying values as if they were min/max. Is that also what you meant? I know that setting an intensity of 2 will be preserved as-is. What I didn't check is whether this would lead to the same behaviour as if the intensity were 1 (which is my guess).

The values vtk gives back aren't clamped if the inputs aren't clamped, whether vtk internally clamps them before rendering or whatever I don't know.

JevinJ commented:
Since everybody else seems busy, just my thoughts on some of your TODOs.

1. I would prefer the 3 args, it's seems the most flexible.

2. It looks like(from my 2 minute look at the source) vtkLight sets the matrix to null if created with a default constructor. If ShallowClone is used and matrix is not null, it's shared. If DeepCopy is used and it's not null, it copies the matrix. world_... makes sense if world space is not the default internally. https://github.com/Kitware/VTK/blob/e9e5ec3cc524123c94a3518462615bcd85deffd2/Rendering/Core/vtkLight.cxx

3. is_on is acceptable to me, the better question is if you want to mirror vtk and switch_on/off with just "switch". is_on looks more readable to me though.

7. I removed it in my testing PR, I can't run plotting and there were no issues so it's not needed.

adeak commented:
Thank you for your thoughts, @JevinJ 

> (1). I would prefer the 3 args, it's seems the most flexible.

Do you mean keeping 3 separate properties for the 3 different colour components? I'd definitely keep that. The question is whether we _also_ want a single "uber" function that sets all the other colours in a single call.
 
> (2). It looks like(from my 2 minute look at the source) vtkLight sets the matrix to null if created with a default constructor. If ShallowClone is used and matrix is not null, it's shared. If DeepCopy is used and it's not null, it copies the matrix. world_... makes sense if world space is not the default internally. https://github.com/Kitware/VTK/blob/e9e5ec3cc524123c94a3518462615bcd85deffd2/Rendering/Core/vtkLight.cxx

OK, I don't see anything that would set the transform matrix to something non-trivial as a side-effect. Although I'll have to see what [these lines](https://github.com/Kitware/VTK/blob/e9e5ec3cc524123c94a3518462615bcd85deffd2/Rendering/Core/vtkLight.cxx#L23) define exactly. But I suspect it's just some declarations and setters/getters, which are otherwise missing from the definitions on that class.

> (3). is_on is acceptable to me, the better question is if you want to mirror vtk and switch_on/off with just "switch". is_on looks more readable to me though.

Similar to the colour case, I already have both `is_on` (settable property), `switch_on/switch_off` as well as `switch` (which can switch on, switch off or toggle). I don't think we've got anything to lose from the variety of options for doing the same thing (which seems to be pyvista's general philosophy). So the only question that remains is the name of the property :)

> (7). I removed it in my testing PR, I can't run plotting and there were no issues so it's not needed.

Good to know, thanks!

JevinJ commented:
> Do you mean keeping 3 separate properties for the 3 different colour components? I'd definitely keep that. The question is whether we also want a single "uber" function that sets all the other colours in a single call.

At the time I meant to keep set_color like vtk has but with 3 args, one for each color type, but the user can set colors in one line with unpacking. We could just copy vtk, but add a bool for setting ambient too and document the behaviour. Maybe rename it set_colors.

akaszynski commented:
I'd like to release the next minor release of `pyvista` soon.  What work still needs to be done for this PR?

adeak commented:
@akaszynski the PR works if we don't mind all the design decisions I'm unsure about (see the above TODOs), I "just" have to add an examples section to showcase the functionality. If this can be omitted for the release then I can finalize the code part of the PR easily.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+
+    @property
+    def positional(self):
+        # TODO: did I understand this correctly?
```
```suggestion
```

Seems correct.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+
+    #### Everything else ####
+
+    def set_color(self, color):
+        # TODO: do we want to remove this?
+        # TODO: do we want instead to set _all_ lights in pyvista?
+        # TODO: do we want to support a three-arg form?
+        """Set the diffuse and specular color of the light.
+
+        Parameters
+        ----------
+        color : tuple(float)
+            The color that should be set for diffuse and specular.
+
+        """
+        self.SetColor(color)
```
```suggestion
```

Completely remove it.  We're creating a new API via this class, and it's only kept within VTK for backwards compatibility, and will still be there with the C++ API.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+        position with respect to it.
+        A scene light is stationary with respect to the scene, as it does not
+        follow the camera. This is the default.
+    """
```
Add an example here
```suggestion
    Examples
    --------
    Create a light at (10, 10, 10) and set its diffuse color to red

    >>> import pyvista as pv
    >>> light = pv.Light()
    >>> light.diffuse_color = 1, 0, 0
    """
```

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+    #### Properties ####
+
+    @property
+    def ambient_color(self):
+        """Return the ambient color of the light."""
+        return self.GetAmbientColor()
+
+    @ambient_color.setter
+    def ambient_color(self, color):
+        """Set the ambient color of the light."""
+        self.SetAmbientColor(color)
```
Here and elsewhere, we need examples within the docstrings for the properties (not necessarily the setter).  I'll add a sphinx autodoc for this class and it will be helpful to have examples for these methods/attributes.

Additionally, `color` should accept a string as in our `plotter` API using `pyvista.plotting.theme.parse_color`.

adeak replied:
Is my understanding correct that setter docstrings don't show up anywhere, and getter docstrings show up in `help(light)`? Hmm, and I guess sphinx might use these too, I'll look that up in the existing docs.

And thanks for the color string suggestion, good catch.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+        """Set the specular color of the light."""
+        self.SetSpecularColor(color)
+
+    # TODO: implement light.color = ... using setattr? no guarantee that a getter would always make sense, so property won't work!
```
```suggestion
```
See other comment for removing the `color` property.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+    def position(self, pos):
+        self.SetPosition(pos)
+
+    @property
+    def world_position(self):
+        # TODO: is this name and configuration OK? Same for world_focal_point
+        # TODO: can a transformation matrix happen accidentally? If not, perhaps we can just not expose these at all!
+        """Return the world space position of the light."""
+        return self.GetTransformedPosition()
```
```suggestion
    @property
    def world_position(self):
        """Return the world space position of the light."""
        return self.GetTransformedPosition()
```

I'm happy with the name, though we should detail that we're returning the value from `GetTransformedPosition` from VTK.  This needs an example as well.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+    def intensity(self, intensity):
+        self.SetIntensity(intensity)
+
+    @property
+    def is_on(self):
+        # TODO: is this name OK? Just "on" sounds too short.
+        """Return whether the light is on."""
+        return bool(self.GetSwitch())
+
+    @is_on.setter
+    def is_on(self, state):
+        """Set whether the light should be on."""
+        self.SetSwitch(state)
```
```suggestion
    @property
    def on(self):
        """Return ``True`` when the light is on."""
        return bool(self.GetSwitch())

    @is_on.setter
    def on(self, state):
        """Turn the light on or off"""
        self.SetSwitch(state)
```

`numpy` has `T` for transpose, so honestly I think we're fine.  The ``is`` is redundant and doesn't make sense in the setter.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+        """Switch off the light."""
+        self.SwitchOff()
+
+    def switch(self, is_on=None):
+        """Set whether the light should be on. Toggles by default.
+
+        Parameters
+        ----------
+        is_on : bool
+            The state of the light to be set (True is on, False is off).
+            Default is to toggle the state.
+
+        """
+        if is_on is None:
+            # toggle
+            is_on = not self.is_on
+        self.SetSwitch(is_on)
```
```suggestion
```
While I like keeping API compatibility with VTK, having both the `on` setter and `switch` seems redundant.  Additionally, were we to keep this, something like `toggle` makes more sense.

For simplicity, lets just remove this and keep just `on`

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+            is_on = not self.is_on
+        self.SetSwitch(is_on)
+
+    def positional_on(self):
+        """Make the light positional.
+
+        Attenuation and cone angles are only used for a positional light.
+
+        """
+        self.PositionalOn()
+
+    def positional_off(self):
+        """Make the light directional.
+
+        Attenuation and cone angles are ignored for a directional light.
+
+        """
+        self.PositionalOff()
```
Would prefer a property/setter rather than duplicating the VTK api.

adeak replied:
I _also_ have a property, so I'll just remove these two since you feel they're redundant.

akaszynski commented on pyvista/plotting/lights.py:
```diff
@@ -0,0 +1,466 @@
+        """
+        self.SetDirectionAngle(elev, azim)
+
+    # TODO: deepcopy?
```
Or just ``copy`` with a ``deep`` bool parameter as in our other classes.

adeak replied:
I have a technical issue with that. I suspect shallow copies should essentially implement [`vtkLight::ShallowClone()`](https://vtk.org/doc/nightly/html/classvtkLight.html#ac7bef205f8a78c3c9081db025e0892f2), but I don't understand vtk enough to know what this would entail for the pyvista class. At for deepcopy I assume it should return a new copy (as per python customs) rather than being a void method on a light object as [`vtkLight::DeepCopy()`](https://vtk.org/doc/nightly/html/classvtkLight.html#abfaf6d69238978415d61841ec6c2ff43). This is why I only poked the deepcopy possibility :)

adeak replied:
OK, I was being confused again. I've got this one working too.

akaszynski commented on pyvista/plotting/plotting.py:
```diff
@@ -3864,6 +3912,39 @@ def reset_camera_clipping_range(self):
+            renderer, see the ``add_light`` method of the Renderer class.
+
+        """
+        # TODO: is this design OK?
```
Yes, adding a light to each render makes sense as this is the default behavior for our other methods.  We still need an example for this method.
```suggestion
```

akaszynski commented on pyvista/plotting/plotting.py:
```diff
@@ -3864,6 +3912,39 @@ def reset_camera_clipping_range(self):
+            is that lights are stripped from every renderer.
+
+        """
+        # TODO: is this design OK?
```
```suggestion
```
Yes.

akaszynski commented on tests/test_lights.py:
```diff
@@ -0,0 +1,219 @@
+import pyvista
+from pyvista.plotting import system_supports_plotting
+# TODO: do we need OFF_SCREEN stuff? And as per tests/test_renderer.py or tests/plotting/test_plotting.py?
+# TODO: do we actually need the @skip_no_plotting decorator when there's no plotter?
```
We do not need the decorator.  We can create lights and VTK/pyvista objects all we want without a framebuffer, it's rendering it that causes a segfault.
```suggestion
```

akaszynski commented:
> I "just" have to add an examples section to showcase the functionality.

An example in our gallery would be great.  Additionally, I'd like a page within the core documentation describing the lighting.  We can keep it very short and include the `autoclass` sphinx directive.  Let me know if you need help adding that.

adeak commented:
Thank you for the detailed review. I'll definitely need help for the sphinx things, but as a first approximation it's enough if you can point me to existing pieces in the code base which are similar, and I can see if I can figure it out.

akaszynski commented:
> point me to existing pieces in the code base which are similar

https://github.com/pyvista/pyvista/blob/master/docs/core/grids.rst

I'd add a lights.rst containing:

```py
Lights
======
Talk about lights and how cool they are and a brief example regarding their creation and usage.  Include a link to the example in your gallery with :ref:`ref_my_light_example` so people can see figures of lights in action.

Brief Example
-------------

.. code:: python

    import pyvista as pv
    light = pv.Light()
    # light usage...

Light API
-------
Description of the ``pyvista.Light`` class.

.. autoclass:: pyvista.Light
   :members:
```

adeak commented:
Technical note: I decided to add a workaround for https://gitlab.kitware.com/vtk/vtk/-/issues/18093 rather than leave our functionality broken (even though `ShadowAttenuation` is probably not a very common property to touch).

I've added a regression test of sorts that verifies that the bug is there in vtk. The idea is that once this is fixed in vtk the test will break, and we can (conditionally?) remove the workaround from the implementation of `copy()`. @akaszynski does this make sense?

adeak commented:
@akaszynski I think I've done all I wanted on the code front. A few remaining uncertainties:
  1. `BasePlotter.lighting` used to refer to the light kit that was the default lighting. This no longer makes any sense, and arguably a single lighting in case of multiple renderers is ambiguous. So I've removed this attribute altogether (nothing broke in our own code at least)
  2. On a related note, there were always 5 lights originally. Even for the "3 lights illumination" case the code repurposed the lights from the light kit, turning two lights off. The current code only uses the number of lights that it needs, which might technically break backwards compatibility too.
  3. Something that I meant to do in a different PR from the start: I think `disable_3_lights()` should be deprecated and renamed `enable_lightkit()`. This is not critical for the release, although it might be better to include the deprecation. If you think this is OK I can add it to this PR.
  
And all that's left are the docs and examples for the website. I have some ideas and the doctests should already provide some ammo, but I'm not sure I'll be able to get it done on the weekend, which isn't great for your release plans. Sorry :(

adeak commented:
@akaszynski this PR is ready for review.

Two questions/remarks:
  1. I couldn't figure out the difference between `.. testcode::` and `.. code::`. I used the former at https://github.com/pyvista/pyvista/pull/1040/files#diff-6859166c6225895e6d3822a8f88ce4417b98f330c7c618aaa1377d3170cb45d3R14 and a few lines below that. It might have to be changed to `.. code::`.
  2. We might want to expand the signature of `pv.Light.__init__`. Currently only the position, focal point, color and light type can be passed on instantiation, because these were the properties I found "fundamental".

akaszynski commented on pyvista/plotting/plotting.py:
```diff
@@ -263,13 +275,21 @@ def __init__(self, shape=(1, 1), border=None, border_color='k',
         self._closed = False
 
+        # lighting style; be forgiving with input (accept underscores and ignore case)
+        if lighting is None:
```
About to suggest to allow `None`.  Nice work here.

akaszynski commented on pyvista/plotting/plotting.py:
```diff
@@ -398,35 +428,47 @@ def _to_pos(elevation, azimuth):
-        for renderer in self.renderers:
-            renderer.RemoveAllLights()
-            self.lighting.AddLightsToRenderer(renderer)
+        intensities = [1, 0.6, 0.5]
+        all_angles = [(45.0, 45.0), (-30.0, -60.0), (-30.0, 60.0)]
+        for intensity, angles in zip(intensities, all_angles):
+            light = pyvista.Light(light_type='camera light')
+            light.intensity = intensity
+            light.position = _to_pos(*angles)
+            for renderer in renderers:
+                renderer.add_light(light)
```
This is *way* cleaner.  I can really see the advantages of having a pythonic lighting class.

akaszynski commented:
Great work!  Examples, implementation, and documentation are great!

I've worked on the unit testing a bit.  In our current testing implementation we have a testing function that compares a cached screenshot with a cached screenshot called ``verify_cache_image`` in `tests/plotting/test_plotting.py`.  I added this a while back so we could verify the renderings we generate are valid, and I thought it would be especially important to have those here.  After all, you could "add" a light without actually adding it to the renderer for some reason.

For example:

```py
@skip_no_plotting
def test_lighting_add_manual_light():
    plotter = pyvista.Plotter(lighting=None)
    plotter.add_mesh(pyvista.Sphere())

    # test manual light addition
    light = pyvista.Light()
    plotter.add_light(light)
    assert plotter.renderer.lights[-1] is light

    plotter.show(before_close_callback=verify_cache_image)
```
![lighting_add_manual_light](https://user-images.githubusercontent.com/11981631/106423239-4c306000-641d-11eb-80db-f3ec7a5aa9fe.png)

The first time you run `verify_cache_image`, it creates the image with no checking.  However, afterwards, `pytest` will always verify that the rendering from the plotter matches the cached image within a certain tolerance.  I thought it would be especially useful for your lighting tests, so I added it in.  Since only one image is created per test, I went ahead and split the tests up.  This helps for regression testing as well.

akaszynski commented:
> I couldn't figure out the difference between `.. testcode::` and `.. code::`. I used the former at https://github.com/pyvista/pyvista/pull/1040/files#diff-6859166c6225895e6d3822a8f88ce4417b98f330c7c618aaa1377d3170cb45d3R14 and a few lines below that. It might have to be changed to `.. code::`.

You guessed right.  `.. testcode::` is actually run during our magical documentation build step and is why the bunny picture actually shows up and you don't have to manually add the image.  This happens in:
https://github.com/pyvista/pyvista/blob/6374c9041daa2c14c08d242e4cc15953c13fa58c/azure-pipelines.yml#L326-L328

> We might want to expand the signature of `pv.Light.__init__`. Currently only the position, focal point, color and light type can be passed on instantiation, because these were the properties I found "fundamental".

I'm satisfied with this at the moment.  This is a non-trivial PR and is really quite good.  I want this merged and then we can add it later.

> BasePlotter.lighting used to refer to the light kit that was the default lighting. This no longer makes any sense, and arguably a single lighting in case of multiple renderers is ambiguous. So I've removed this attribute altogether (nothing broke in our own code at least)

None of the plotting tests failed, and a quick visual inspection shows the lighting behavior hasn't changed.  I'm fine with this.

> On a related note, there were always 5 lights originally. Even for the "3 lights illumination" case the code repurposed the lights from the light kit, turning two lights off. The current code only uses the number of lights that it needs, which might technically break backwards compatibility too.

Doesn't seem to.  A side by side comparison with `master` shows 0 variation.  Plus your code implementation is ultra concise.  Much less to maintain.

> Something that I meant to do in a different PR from the start: I think disable_3_lights() should be deprecated and renamed enable_lightkit(). This is not critical for the release, although it might be better to include the deprecation. If you think this is OK I can add it to this PR.

Yes.  When working on the testing I had no idea why this was even included.  I'll depreciate it and finalize this PR.

akaszynski reviewed (approved):
Again, nice work.  Lighting is something I set and forget because it wasn't implemented pythonically, and I'll be able to make great use of this.  

Really liked the examples as well:
![image](https://user-images.githubusercontent.com/11981631/106425754-9c112600-6421-11eb-8410-9ca6ee882fc5.png)

adeak commented:
Thanks, @akaszynski! Also for the feedback.

>  In our current testing implementation we have a testing function that compares a cached screenshot with a cached screenshot called verify_cache_image in tests/plotting/test_plotting.py

I saw that when the puppy texture broke. It just slipped my mind while I was working on this PR. Thanks for the explanation of how it works. The functionality seems broken right now, for which I'll open an issue :)

> why the bunny picture actually shows up and you don't have to manually add the image

I did manually add the image, though with a `.. image` directive :)

---

I found a few minor issues that need fixing up, so I opened a PR with these: https://github.com/pyvista/pyvista/pull/1131
